File: Graph Structures - Incidence List.cpp
#include <bits/stdc++.h>
using namespace std;
template <bool Directed, typename TEdge, bool Index> struct graph {
    using EType = TEdge;
    vector<TEdge> edges;
    vector<vector<int>> adj;
    graph(int n) : adj(n + Index) {}
    graph(int n, int m) : graph(n) { edges.reserve(m << not Directed); }
    TEdge& operator()(int e) { return edges[e]; }
    vector<int>& operator[](int u) { return adj[u]; }
    int size() { return adj.size() - Index; }
    void append(int u, const TEdge& e) {
        adj[u].push_back(edges.size());
        edges.push_back(e);
    }
    void add_edge(const TEdge& e) {
        append(e.u, e);
        if(not Directed) append(e.v, e.reverse());
    }
};

File: Graph Structures - Incidence List.py
class graph:
    def __init__(self, n, oneway, index=1):
        self.oneway, self.index, self.adj, self.edges = oneway, index, [[] for i in range(n + index)], []
    def __len__(self): return len(self.adj) - self.index
    def __getitem__(self, u): return self.adj[u]
    def __call__(self, e): return self.edges[e]
    def __repr__(self): return repr(self.edges)
    def append(self, u, edge): self.adj[u].append(len(self.edges)); self.edges.append(edge)
    def add_edge(self, edge):
        self.append(edge.u, edge)
        if not self.oneway: self.append(edge.v, reversed(edge))
